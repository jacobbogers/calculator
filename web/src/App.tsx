/* vender */
import React, { useCallback, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import classNames from 'classnames';

/* styles */
import styles from './App.module.scss';

/* components */
import CalculatorBtn from './components/CalculatorBtn';
import KeyBoard from './components/KeyBoard';
import Display from './components/Display';

/* redux */
import type { StoreLayout } from './lib/reducers';

/* misc */
import { createAction } from './lib/actions';

export default function App() {
    const dispatch = useDispatch();
    const isDirty = useSelector<{ tainted: boolean }, boolean>((state) => state.tainted);
    const networkRequestInTransit = useSelector<StoreLayout, boolean>((state) => state.inTransit);

    const safeDispatch = useCallback(
        (...args: Parameters<typeof dispatch>) => {
            if (networkRequestInTransit) {
                return;
            }
            return dispatch(...args);
        },
        [dispatch, networkRequestInTransit]
    );

    const className = useMemo(
        () =>
            classNames({
                [styles.app]: true,
                [styles.wait]: networkRequestInTransit
            }),
        [networkRequestInTransit]
    );

    return (
        <div className={className}>
            <KeyBoard>
                <Display />
                {/* row 1 */}
                <CalculatorBtn disabled darker>
                    Rad
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    Deg
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    x!
                </CalculatorBtn>

                <CalculatorBtn disabled darker>
                    (
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    )
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    %
                </CalculatorBtn>
                {isDirty ? (
                    <CalculatorBtn
                        data-testid="btnClearEntry"
                        darker
                        smaller
                        onClick={() => safeDispatch(createAction('clearEntry'))}
                    >
                        CE
                    </CalculatorBtn>
                ) : (
                    <CalculatorBtn
                        data-testid="btnClearAll"
                        darker
                        smaller
                        onClick={() => safeDispatch(createAction('clearAll'))}
                    >
                        AC
                    </CalculatorBtn>
                )}
                {/* row 2 */}

                <CalculatorBtn disabled darker>
                    Inv
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    sin
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    ln
                </CalculatorBtn>

                <CalculatorBtn data-testid="btn7" onClick={() => safeDispatch(createAction('digit', '7'))}>
                    7
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn8" onClick={() => safeDispatch(createAction('digit', '8'))}>
                    8
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn9" onClick={() => safeDispatch(createAction('digit', '9'))}>
                    9
                </CalculatorBtn>
                <CalculatorBtn
                    data-testid="btnDivide"
                    darker
                    larger
                    onClick={() => safeDispatch(createAction('divide'))}
                >
                    ÷
                </CalculatorBtn>

                {/* row 3 */}

                <CalculatorBtn disabled darker>
                    π
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    cos
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    log
                </CalculatorBtn>

                <CalculatorBtn data-testid="btn4" onClick={() => safeDispatch(createAction('digit', '4'))}>
                    4
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn5" onClick={() => safeDispatch(createAction('digit', '5'))}>
                    5
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn6" onClick={() => safeDispatch(createAction('digit', '6'))}>
                    6
                </CalculatorBtn>
                <CalculatorBtn
                    data-testid="btnMultiply"
                    darker
                    larger
                    onClick={() => safeDispatch(createAction('multiply'))}
                >
                    ×
                </CalculatorBtn>

                {/* row 4 */}

                <CalculatorBtn disabled darker>
                    e
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    tan
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    √
                </CalculatorBtn>

                <CalculatorBtn data-testid="btn1" onClick={() => safeDispatch(createAction('digit', '1'))}>
                    1
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn2" onClick={() => safeDispatch(createAction('digit', '2'))}>
                    2
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn3" onClick={() => safeDispatch(createAction('digit', '3'))}>
                    3
                </CalculatorBtn>
                <CalculatorBtn
                    data-testid="btnSubtract"
                    darker
                    larger
                    onClick={() => safeDispatch(createAction('subtract'))}
                >
                    −
                </CalculatorBtn>

                {/* row 5 */}

                <CalculatorBtn disabled darker>
                    Ans
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    EXP
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    <span>xʸ</span>
                </CalculatorBtn>

                <CalculatorBtn data-testid="btn0" onClick={() => safeDispatch(createAction('digit', '0'))}>
                    0
                </CalculatorBtn>
                <CalculatorBtn data-testid="btnDot" bolder onClick={() => safeDispatch(createAction('digit', '.'))}>
                    .
                </CalculatorBtn>
                <CalculatorBtn
                    data-testid="btnSubmit"
                    bolder
                    larger
                    submit
                    onClick={() => safeDispatch(createAction('result'))}
                >
                    =
                </CalculatorBtn>
                <CalculatorBtn data-testid="btnAdd" darker larger onClick={() => safeDispatch(createAction('add'))}>
                    +
                </CalculatorBtn>
            </KeyBoard>
        </div>
    );
}
