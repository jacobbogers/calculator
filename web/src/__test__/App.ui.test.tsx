import React from 'react';
import createStore from '../lib/store';
import App from '../App';

import { customRender } from './test-helper';

import { describe, it, expect } from 'vitest';

describe('User Interface', () => {
    it('check initial DOM structure', () => {
        const store = createStore();
        const { unmount } = customRender(<App />, store);
        expect(document.body).toMatchSnapshot();
        unmount();
    });
});
