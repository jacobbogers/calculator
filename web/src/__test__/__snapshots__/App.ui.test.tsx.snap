// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User Interface > check initial DOM structure 1`] = `
<body>
  <div>
    <div
      class="_app_ce1e69"
    >
      <div
        class="_keyboard_0b2576"
      >
        <div
          class="_display_1250f0"
          tabindex="1"
        >
          <div
            class="_theGrid_1250f0"
          >
            <div
              class="_history_0ea9b2"
            >
              <svg
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13 3a9 9 0 0 0-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42A8.954 8.954 0 0 0 13 21a9 9 0 0 0 0-18zm-1 5v5l4.28 2.54.72-1.21-3.5-2.08V8H12z"
                />
              </svg>
            </div>
            <div
              class="_last_history_9cba2d"
            >
              <span
                data-testid="lastHistoryLine"
              />
            </div>
            <div
              class="_command_line_ab67cd"
            >
              <div>
                <span
                  data-testid="commandLine"
                >
                  0
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          Rad
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          Deg
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          x!
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          (
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          )
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          %
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _smaller_0e6df9"
          data-testid="btnClearAll"
        >
          AC
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          Inv
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          sin
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          ln
        </div>
        <div
          class="_btn_0e6df9"
          data-testid="btn7"
        >
          7
        </div>
        <div
          class="_btn_0e6df9"
          data-testid="btn8"
        >
          8
        </div>
        <div
          class="_btn_0e6df9"
          data-testid="btn9"
        >
          9
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _larger_0e6df9"
          data-testid="btnDivide"
        >
          ÷
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          π
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          cos
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          log
        </div>
        <div
          class="_btn_0e6df9"
          data-testid="btn4"
        >
          4
        </div>
        <div
          class="_btn_0e6df9"
          data-testid="btn5"
        >
          5
        </div>
        <div
          class="_btn_0e6df9"
          data-testid="btn6"
        >
          6
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _larger_0e6df9"
          data-testid="btnMultiply"
        >
          ×
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          e
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          tan
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          √
        </div>
        <div
          class="_btn_0e6df9"
          data-testid="btn1"
        >
          1
        </div>
        <div
          class="_btn_0e6df9"
          data-testid="btn2"
        >
          2
        </div>
        <div
          class="_btn_0e6df9"
          data-testid="btn3"
        >
          3
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _larger_0e6df9"
          data-testid="btnSubtract"
        >
          −
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          Ans
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          EXP
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _disabled_0e6df9"
        >
          <span>
            xʸ
          </span>
        </div>
        <div
          class="_btn_0e6df9"
          data-testid="btn0"
        >
          0
        </div>
        <div
          class="_btn_0e6df9 _bolder_0e6df9"
          data-testid="btnDot"
        >
          .
        </div>
        <div
          class="_btn_0e6df9 _bolder_0e6df9 _larger_0e6df9 _submit_0e6df9"
          data-testid="btnSubmit"
        >
          =
        </div>
        <div
          class="_btn_0e6df9 _darker_0e6df9 _larger_0e6df9"
          data-testid="btnAdd"
        >
          +
        </div>
      </div>
    </div>
  </div>
</body>
`;
